core/templates/services/html-length.service.ts

```diff
@@ -50,51 +65,129 @@ export class HtmlLengthService {
     }
     const sanitizedHtml = this.sanitizer.sanitize(
       SecurityContext.HTML, htmlString) as string;
-    let totalWeight = this.calculateBaselineLength(sanitizedHtml);
+    // Identify custom tags using regex on the original HTML string.
+    const customTags = htmlString.match(CUSTOM_TAG_REGEX);
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    /* CustomTagsLength is being passed to calculateBaselineLength()
+    method to ensure try-catch block should not raise error if
+    we use non-interactive RTE */
+    let totalWords = this.calculateBaselineLength(
+      sanitizedHtml, CALCULATION_TYPE_WORD, customTagsLength);
+
+    if (customTags) {
+      for (const customTag of customTags) {
+        totalWords += this.getWeightForNonTextNodes(
+          customTag, CALCULATION_TYPE_WORD);
+      }
+    }
+    return totalWords;
+  }
+
+  computeHtmlLengthInCharacters(htmlString: string): number {
+    if (!htmlString) {
+      this.loggerService.error('Empty string was passed to compute length');
+      return 0;
+    }
 
+    const sanitizedHtml = this.sanitizer.sanitize(
+      SecurityContext.HTML, htmlString) as string;
     // Identify custom tags using regex on the original HTML string.
-    const customTagRegex = /<oppia-noninteractive-(?:math|image)[^>]*>/g;
-    const customTags = htmlString.match(customTagRegex);
+    const customTags = htmlString.match(CUSTOM_TAG_REGEX);
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    /* CustomTagsLength is being passed to calculateBaselineLength()
+    method to ensure try-catch block should not raise error if
+    we use non-interactive RTE */
+    let totalCharacters = this.calculateBaselineLength(
+      sanitizedHtml, CALCULATION_TYPE_CHARACTER, customTagsLength);
 
     if (customTags) {
       for (const customTag of customTags) {
-        totalWeight += this.getWeightForNonTextNodes(customTag);
+        totalCharacters += this.getWeightForNonTextNodes(
+          customTag, CALCULATION_TYPE_CHARACTER);
       }
     }
-    return totalWeight;
+    return totalCharacters;
   }
 
-  private calculateBaselineLength(sanitizedHtml: string): number {
+  calculateBaselineLength(
+      sanitizedHtml: string,
+      calculationType: CalculationType, CustomTagsLength: number): number {
     let domparser = new DOMParser();
-    let dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+    let dom: Document;
+    try {
+      dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+      if (dom.body.children.length === 0 && !CustomTagsLength) {
+        throw new Error(
```

Comment:
We need this case to raise an error if normal string is been passed, if we remove this, it makes debugging harder for future folks.

Reply:
Well ... this was when you told me in the debugging doc that "I analysed method computeHtmlLengthInCharacters() receives html tags as a string which it parse later on and we were giving it a normal string so there was nothing to parse." So I assumed that the method could inherently only take HTML strings.

But from this conversation it sounds like that's not a requirement and we can regard a regular string as an HTML string too. So I don't think you need the error and I think this method should be able to process strings with no html tags too (you should add tests for that). I also would suggest adding some JSDoc to your methods so that the contract is clear and there is no confusion. 

As discussed before, this also probably means you can remove the CustomTagsLength argument.